tfhe/src/safe_serialization.rs

```diff
@@ -0,0 +1,111 @@
+use bincode::Options;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+const SERIALIZATION_VERSION: &str = "0.1";
+
+const VERSION_LENGTH_LIMIT: u64 = 100;
+
+const TYPE_NAME_LENGTH_LIMIT: u64 = 1000;
+
+pub fn safe_serialize<T: Serialize + Named>(
+    object: &T,
+    mut writer: impl std::io::Write,
+    serialized_size_limit: u64,
+) -> bincode::Result<()> {
+    let options = bincode::DefaultOptions::new()
+        .with_fixint_encoding()
+        .with_limit(0);
+
+    options
+        .with_limit(VERSION_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &SERIALIZATION_VERSION.to_owned())?;
+
+    options
+        .with_limit(TYPE_NAME_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &T::NAME.to_owned())?;
+
+    options
+        .with_limit(serialized_size_limit)
+        .serialize_into(&mut writer, object)?;
+
+    Ok(())
+}
+
+pub fn safe_deserialize<T: DeserializeOwned + Named>(
+    mut reader: impl std::io::Read,
+    serialized_size_limit: u64,
+) -> Result<T, String> {
+    let options = bincode::DefaultOptions::new()
+        .with_fixint_encoding()
+        .with_limit(0);
+
+    let deserialized_version: String = options
+        .with_limit(VERSION_LENGTH_LIMIT)
+        .deserialize_from::<_, String>(&mut reader)
+        .map_err(|err| err.to_string())?;
+
+    if deserialized_version != SERIALIZATION_VERSION {
+        return Err(format!(
+            "Expected version {}, got version {}",
+            SERIALIZATION_VERSION, deserialized_version
+        ));
+    }
+
+    let deserialized_type: String = options
+        .with_limit(TYPE_NAME_LENGTH_LIMIT)
+        .deserialize_from::<_, String>(&mut reader)
+        .map_err(|err| err.to_string())?;
+
+    if deserialized_type != T::NAME {
+        return Err(format!(
+            "Expected type {}, got type {}",
+            T::NAME,
+            deserialized_type
+        ));
+    }
+
+    options
+        .with_limit(serialized_size_limit)
+        .deserialize_from(&mut reader)
+        .map_err(|err| err.to_string())
+}
+
+pub trait ParameterSetConformant {
+    type ParameterSet;
+
+    fn is_conformant(&self, param: &Self::ParameterSet) -> bool;
+}
+
+pub fn safe_deserialize_conformant<T: DeserializeOwned + Named + ParameterSetConformant>(
```

Comment:
consider adding #[track_caller] so that the panic/error shows where in the calling code this error is originating

Reply:
The function is supposed to return errors, not `panic`
Isn't `track_caller` mostly useful on `panic`s?

Reply:
ah yes my bad, don't know why I thought we might panic in there